paywall/src/data-iframe/blockchainHandler/purchaseKey.js

```diff
@@ -49,34 +50,44 @@ export async function processKeyPurchaseTransactions({
 }) {
   let transactions = startingTransactions
   let key = startingKey
+  let linkedKey
   let result
   walletService.addListener('transaction.pending', walletAction)
-  const afterEventProcessed = () => {
-    if (transactions !== result.transactions) {
-      transactions = result.transactions
-      key = result.key
-      update(transactions, key)
+  try {
+    const afterEventProcessed = () => {
+      if (transactions !== result.transactions) {
+        transactions = result.transactions
+        key = result.key
+        update(transactions, key)
+      }
     }
-  }
-
-  result = await submittedListener({
-    lockAddress,
-    existingTransactions: transactions,
-    existingKey: key,
-    walletService,
-    requiredConfirmations,
-  })
-  afterEventProcessed()
 
-  do {
-    result = await updateListener({
+    result = await submittedListener({
       lockAddress,
       existingTransactions: transactions,
       existingKey: key,
+      walletService,
       web3Service,
       requiredConfirmations,
     })
     afterEventProcessed()
-  } while (key.status === 'confirming')
-  walletService.removeListener('transaction.pending', walletAction)
+
+    do {
+      result = await updateListener({
+        lockAddress,
+        existingTransactions: transactions,
+        existingKey: key,
+        web3Service,
+        requiredConfirmations,
+      })
+      afterEventProcessed()
+      linkedKey = linkTransactionsToKey({
+        key,
+        transactions,
+        requiredConfirmations,
+      })
+    } while (linkedKey.status === 'confirming')
+  } finally {
```

Comment:
this is basically the same, but wrapped in a try/finally so we don't leak memory on errors thrown